### ISSUE TITLE
Co-locating images together with a blog post in a folder

### ISSUE THREAD

**Issue by zachleat** [11ty/eleventy#2428]
Via:
* @AleksandrHovhannisyan https://twitter.com/hovhaDovah/status/1525086605303857152 and https://twitter.com/hovhaDovah/status/1525271121549004804
* @SaraSoueidan https://twitter.com/SaraSoueidan/status/1533468684089536513

```
I want to co-locate posts and images like this:

yyyy-mm-dd-slug/index.md
yyyy-mm-dd-slug/images/image-files-go-here

Use Markdown images natively:

![Alt](./images/the-image.png)

And replace them with 11ty image output.
```

Just want to make sure this captures the workflow y’all would like to see!

**Comment by AleksandrHovhannisyan** [comment]
Thanks! I did get the first part working after my PR got merged (co-locating images), but I don't think auto-converting Markdown images to 11ty images is currently possible. I got very close, but I ultimately ran into the gotcha noted here for `addExtension`: https://twitter.com/hovhaDovah/status/1525271580070248448. Basically, it opts me out of all of the processing that 11ty already does for Markdown files, so I would then be responsible for handling includes, Liquid, and everything else that automagically goes on behind the scenes.

I think it would be cool if 11ty exposed some kind of post-processing API that allows you to query all files matching a format/extension (like `addExtension`) and modify the output, before 11ty writes that file to the build output. This would allow me to leverage 11ty's initial processing for my blog posts and then later go in and correct the images, replacing them with 11ty images as Ben Holmes demonstrated here: https://github.com/Holben888/11ty-image-optimization-demo. What do you think?

**Comment by SaraSoueidan** [comment]
A little late to the comments, but YES this is exactly what I had in mind. 🙏🏻

**Comment by aogriffiths** [comment]
This is exactly what I would like too 🙏🏻 makes it soo much easier to manager pages that have a lot of images, e.g. tutorials that have screenshots.

**Comment by aogriffiths** [comment]
this seems close but not using standard markdown https://gfscott.com/blog/eleventy-img-without-central-image-directory/ and here is a seemingly simple solution https://scottwhittaker.net/posts/colocate-images-with-post-using-eleventy

**Comment by aaaaargZombies** [comment]
Just wanted to add a big vote in support of doing this. I think it would be one of the main barriers preventing my workplace moving from hugo to eleventy. We do a bunch of stuff with community groups and digital autonomy, one of the things we've found when helping set up a static site for a group and on boarding them is that the idea of single piece of work being split up and spread across different directories (and those directories contain pieces of other bits of work!?) is really not intuitive and can be quite overwhelming.

**Comment by davidvasandani** [comment]
I accomplished the images in a posts directory [here](https://github.com/davidvasandani/blog/blob/main/.eleventy.config.images.js).

**Comment by AleksandrHovhannisyan** [comment]
@zachleat Am I correct in understanding that this issue will be resolved with the new `@11ty/eleventy-img@4.0+` `eleventyImageTransformPlugin`? I did a quick little test and it seems to work as expected with relative image paths and plain-old Markdown images (i.e., `![alt](./src)`).

**Comment by ptsefton** [comment]
@AleksandrHovhannisyan I tried that plugin and it did not work out very well for me. It DOES result in images co-located with the page in the output, but it rewrites the images and changes their file names and running it on my 20 year old blog it was throwing lots of errors, complaining about images which work fine but which it did not like for some reason.  think it's changing the resolution and size of the images as well, which might be useful in some situations, but all I was after and what brought me to this issue was that I wanted to just copy source images from the input to the output.

 In the end I added a hacky thing to my 11tyconfig to copy stuff - complete with a catch block that does no error reporting.

```
const Path = require("Path")


module.exports = function (eleventyConfig) {



  eleventyConfig.addTransform("copy", function (content) {
    if ((this.page.outputPath || "").endsWith(".html")) {

      const fromDir = Path.dirname(this.page.inputPath);
      const toDir = Path.dirname(this.page.outputPath);
      console.log("FROM", fromDir);
      console.log("TO", toDir);
      for (let f of fs.readdirSync(fromDir)) {
        if (!f.endsWith(".md")) {
          try {
            if (!fs.existsSync(Path.join(toDir, f))) {
              console.log("copying", f, fromDir, toDir)
              fs.copyFileSync(Path.join(fromDir, f), Path.join(toDir, f))
            }
          } catch (error) {

          }
        }
      }

      // Return content as-is
      return content;
    });

};


```

**Comment by tettoffensive** [comment]
@zachleat I'm a little confused. Does this just work out of the box now with 11ty 3.0? I'm trying to configure each of my markdown posts in a directory with all the images for that post in the same subdirectory. Each post has a heroImage which I defined in the front-matter and will also have other images referenced in the markdown file. It's a little hard to parse what's going on in the commit @AleksandrHovhannisyan posted. 

(I'm new to 11ty and have a fresh repository)

**Comment by AleksandrHovhannisyan** [comment]
> @zachleat I'm a little confused. Does this just work out of the box now with 11ty 3.0? I'm trying to configure each of my markdown posts in a directory with all the images for that post in the same subdirectory. Each post has a heroImage which I defined in the front-matter and will also have other images referenced in the markdown file. It's a little hard to parse what's going on in the commit @AleksandrHovhannisyan posted.
>
> (I'm new to 11ty and have a fresh repository)

@tettoffensive It should work in 3.0. My setup boils down to:

```
src/_posts
├── 2024-10-17-my-post-slug
│   ├── images
│   │   ├── image1.jpg
│   │   └── image2.jpg
│   └── index.md
```

Each Markdown post lives in a folder in `src/_posts/`. The content for each post goes in `index.md` and the images could either be flattened at the same level or in their own directory.

Then, using the new [Eleventy image transforms plugin](https://www.11ty.dev/docs/plugins/image/#eleventy-transform), I reference local images using relative paths, like so:

```md
---
front: matter
---

Hello from index.md

![alt text](./images/image1.jpg)
```

And the transform plugin will write the output images to the post's permalink directory in your dist folder.

**Comment by tettoffensive** [comment]
@AleksandrHovhannisyan Thanks! That works, but I also have a cover image for each post so in the front matter I have:

```
---
title: 'Blog post title'
description: 'Lorem ipsum dolor sit amet'
coverImage: './cover.png'
---
```

Then that image would show up on both the index page and at the top of each post page in the template. It seems like I need to do something to get that set up as it's not working out of the box.

Or maybe there's another way to accomplish this?

**Comment by AleksandrHovhannisyan** [comment]
@tettoffensive Yes, you'll need a custom solution for local thumbnails/cover images. What I do on my site is I have a JavaScript data file in my `_posts` directory that derives some data from each post's front matter:

https://github.com/AleksandrHovhannisyan/aleksandrhovhannisyan.com/blob/f014abe5b04c2bf77704b154aff184a3fb9d5e46/packages/web/src/_posts/_posts.11tydata.js#L28-L45

What I basically do is:

1. Check if the image points to a remote src. If it does, we're done.
2. If it doesn't, I assume it points to a local image, like `./images/thumbnail.png`.
3. Convert the local path to a root-relative path using Node.js's `path.parse` and `path.join`.
4. Get the path to the output/dist folder where I want to save the image.
5. Pass the image `src` and `outputDir` to a custom function that uses the Eleventy image plugin to parse the image metadata and returns the image URL.

It's a bit complicated but I'm not aware of any existing workarounds.

**Comment by tettoffensive** [comment]
@AleksandrHovhannisyan Your solution definitely makes sense for OpenGraph Images and but I noticed I was able to use the new image transform plugin and just reference the image in my layout template like so and it seemed to work fine:

```html
<img src="{{ coverImage }}" alt="..." />
```

**Comment by AleksandrHovhannisyan** [comment]
@tettoffensive I believe you'd want to use `post.data.coverImage` in that example, not `post.coverImage` (unless this was a typo)

**Comment by tettoffensive** [comment]
Ah yes. It should be `<img src="{{ post.url }}{{ post.data.coverImage }}" alt="">`

**Comment by zachleat** [comment]
This will be solved by the AutoCopy plugin and a PR for this (with a few preliminary docs) is here: https://github.com/11ty/eleventy/pull/3573

**Comment by zachleat** [comment]
Let’s group on this one over at #3552 3552
